runtime/src/bank/executor_cache.rs

```diff
@@ -0,0 +1,377 @@
+        self.executors.read()
+    }
+    pub(super) fn write(&mut self) -> LockResult<RwLockWriteGuard<CachedExecutors>> {
+        if self.shared {
+            self.shared = false;
+            let local_cache = (*self.executors.read().unwrap()).clone();
+            self.executors = Arc::new(RwLock::new(local_cache));
+        }
+        self.executors.write()
+    }
+}
+
+pub(super) const MAX_CACHED_EXECUTORS: usize = 100; // 10 MB assuming programs are around 100k
+
+/// LFU Cache of executors
+#[derive(Debug)]
+pub(super) struct CachedExecutors {
+    max: usize,
+    executors: HashMap<Pubkey, (AtomicU64, Arc<dyn Executor>)>,
+}
+impl Default for CachedExecutors {
+    fn default() -> Self {
+        Self {
+            max: MAX_CACHED_EXECUTORS,
+            executors: HashMap::new(),
+        }
+    }
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl AbiExample for CachedExecutors {
+    fn example() -> Self {
+        // Delegate AbiExample impl to Default before going deep and stuck with
+        // not easily impl-able Arc<dyn Executor> due to rust's coherence issue
+        // This is safe because CachedExecutors isn't serializable by definition.
+        Self::default()
+    }
+}
+
+impl Clone for CachedExecutors {
+    fn clone(&self) -> Self {
+        let mut executors = HashMap::new();
+        for (key, (count, executor)) in self.executors.iter() {
+            executors.insert(
+                *key,
+                (AtomicU64::new(count.load(Relaxed)), executor.clone()),
+            );
+        }
+        Self {
+            max: self.max,
+            executors,
+        }
+    }
+}
+impl CachedExecutors {
+    pub(super) fn new(max: usize) -> Self {
+        Self {
+            max,
+            executors: HashMap::new(),
+        }
+    }
+    pub(super) fn get(&self, pubkey: &Pubkey) -> Option<Arc<dyn Executor>> {
+        self.executors.get(pubkey).map(|(count, executor)| {
+            count.fetch_add(1, Relaxed);
+            executor.clone()
+        })
+    }
+    pub(super) fn put(&mut self, pubkey: &Pubkey, executor: Arc<dyn Executor>) {
+        if !self.executors.contains_key(pubkey) && self.executors.len() >= self.max {
+            let mut least = u64::MAX;
+            let default_key = Pubkey::default();
+            let mut least_key = &default_key;
+            for (key, (count, _)) in self.executors.iter() {
+                let count = count.load(Relaxed);
+                if count < least {
+                    least = count;
+                    least_key = key;
+                }
+            }
+            let least_key = *least_key;
+            let _ = self.executors.remove(&least_key);
+        }
+        let _ = self
+            .executors
+            .insert(*pubkey, (AtomicU64::new(0), executor));
+    }
+    pub(super) fn remove(&mut self, pubkey: &Pubkey) {
+        let _ = self.executors.remove(pubkey);
+    }
+}
+
+impl Bank {
```

Comment:
Hmm why is this? 

Typically I don't like that as well, but given the Bank impl is pretty big, but can be refactored into logical groups that makes it easier to read and cross reference, I thought in this case it would be suitable.